### PR TITLE
Optimized representation of Infos.

### PR DESCRIPTION
**Pull request by sjrd**
~~Based on #4811. Only the last commit belongs to this PR. Otherwise ready to review.~~

### CODE REVIEW

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala [hunk 1]
```diff
@@ -354,51 +356,116 @@ object Infos {
...
+          staticFieldsWritten = toLikelyEmptyList(staticFieldsWritten),
+          methodsCalled = toLikelyEmptyList(methodsCalled),
+          methodsCalledStatically = toLikelyEmptyList(methodsCalledStatically),
+          methodsCalledDynamicImport = toLikelyEmptyList(methodsCalledDynamicImport),
+          jsNativeMembersUsed = toLikelyEmptyList(jsNativeMembersUsed),
+          flags = flags
       )
     }
+
+    private def toLikelyEmptyList[A](set: mutable.Set[A]): List[A] =
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala:466]: Does this change the execution time? At least for 2.13, the `toList` call seems to end up here (with `Nil` as receiver):

https://github.com/scala/scala/blob/abb224a35fa7c386671eb7fb941e870bde0ec447/src/library/scala/collection/immutable/List.scala#L146-L165

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala [hunk 2]
```diff
@@ -1316,139 +1316,120 @@ private final class Analyzer(config: CommonPhaseConfig,
...
+        }
 
-    for (className <- data.staticallyReferencedClasses) {
-      staticDependencies += className
-      lookupClass(className)(_ => ())
-    }
+        /* Since many of the lists below are likely to be empty, we always
+         * test `!list.isEmpty` before calling `foreach` or any other
+         * processing, avoiding closure allocations.
+         */
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala:1353]: Does that make sense to do it here for the calls that allocate closures further down? (e.g. `readFields`)? Feels like we should adjust the definition sites instead?

**Review by gzm0 — APPROVED** [review]
Only nits.

Out of curiosity: Do we know how this affects memory numbers?

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala (hunk 1 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala:466]: In 2.12 that goes to `TraversableOnce.toList`, which then goes to `TraversableLike.to[List]`:
https://github.com/scala/scala/blob/e66a4617c980314dd1c47c68077c7f8ae2e4183e/src/library/scala/collection/TraversableLike.scala#L782-L788
which is not great. It invokes a `CanBuildFrom`, creates a builder, and sends itself to it with `++= this`.

I can't really test whether it changes the execution time right now. I'm on Windows during the week-end, and timings are a lot less reliable.

Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala:466]: Ah, yeah, fair enough. Then let's leave it.

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala (hunk 2 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala:1353]: Meh. It's a trade-off. It only applies to `readFields` and `writeFields`, really. If I move the `isEmpty` check to them, and I read this piece of code, I wonder why those *don't* have the `isEmpty` check. It seems more systematic to do it for every list here, and not wonder.

**Comment by sjrd** [comment]
> Out of curiosity: Do we know how this affects memory numbers?

No, I didn't check memory usage. I think it shouldn't be very different. We have more memory due to the empty list slots, but we also avoid hash map entries when we reach at least 2 things for the same class.

---

Given your comment at https://github.com/scala-js/scala-js/pull/4813#pullrequestreview-1314517501, I rebased only the relevant commit onto `main`, decoupling from the contents of #4811. But otherwise I did not change anything.

**Review by gzm0 — APPROVED** [review]

**Pull request merged**